exporter/exporterhelper/internal/bounded_memory_queue.go

```diff
@@ -17,85 +17,144 @@ import (
 // where the queue is bounded and if it fills up due to slow consumers, the new items written by
 // the producer are dropped.
 type boundedMemoryQueue struct {
-	stopWG       sync.WaitGroup
-	size         *atomic.Uint32
-	stopped      *atomic.Bool
+	size         int
+	eventChan    chan any
 	items        chan Request
-	capacity     uint32
+	capacity     int
 	numConsumers int
+	stopped      *atomic.Bool
+	overflow     *atomic.Bool
 }
 
 // NewBoundedMemoryQueue constructs the new queue of specified capacity, and with an optional
 // callback for dropped items (e.g. useful to emit metrics).
 func NewBoundedMemoryQueue(capacity int, numConsumers int) ProducerConsumerQueue {
 	return &boundedMemoryQueue{
 		items:        make(chan Request, capacity),
-		stopped:      &atomic.Bool{},
-		size:         &atomic.Uint32{},
-		capacity:     uint32(capacity),
+		eventChan:    make(chan any),
+		capacity:     capacity,
 		numConsumers: numConsumers,
+		stopped:      &atomic.Bool{},
+		overflow: func() *atomic.Bool {
+			o := &atomic.Bool{}
+			if capacity == 0 {
+				o.Store(true)
+			}
+			return o
+		}(),
 	}
 }
 
-// StartConsumers starts a given number of goroutines consuming items from the queue
+// Start starts a given number of goroutines consuming items from the queue
 // and passing them into the consumer callback.
 func (q *boundedMemoryQueue) Start(_ context.Context, _ component.Host, set QueueSettings) error {
 	var startWG sync.WaitGroup
+	startWG.Add(1)
+	go func() {
+		startWG.Done()
+		q.eventLoop()
+	}()
+
 	for i := 0; i < q.numConsumers; i++ {
-		q.stopWG.Add(1)
 		startWG.Add(1)
 		go func() {
 			startWG.Done()
-			defer q.stopWG.Done()
 			for item := range q.items {
-				q.size.Add(^uint32(0))
 				set.Callback(item)
+				q.eventChan <- true
 			}
 		}()
 	}
 	startWG.Wait()
 	return nil
 }
 
-// Produce is used by the producer to submit new item to the queue. Returns false in case of queue overflow.
-func (q *boundedMemoryQueue) Produce(item Request) bool {
-	if q.stopped.Load() {
-		return false
-	}
+func (q *boundedMemoryQueue) eventLoop() {
+	overflow := q.capacity == 0
+	var exitChan chan struct{}
+	for {
+		e := <-q.eventChan
+		if req, ok := e.(Request); ok {
+			if overflow {
+				continue
+			}
+			q.size++
+			q.items <- req
+			overflow = q.capacity == 0 || q.size >= q.capacity
+			if overflow {
+				q.overflow.Store(true)
+			}
+			continue
+		}
+		if done, ok := e.(bool); ok && done {
+			q.size--
+			if exitChan != nil && q.size == 0 {
+				break
+			}
+			previousOverflow := overflow
+			overflow = q.capacity == 0 || q.size >= q.capacity
+			if overflow != previousOverflow {
+				q.overflow.Store(false)
+			}
```

Comment:
I think you can use `atomic.CompareAndSwap` here to make it easier?

Reply:
No, because `overflow` is actually a local loop variable, which is only touched by one thread since it's only accessed by the channel events. The `q.overflow` is an atomic boolean and requires a lock for updates.